### PR TITLE
[Core] Raylet to pick the node manager port

### LINKED ISSUE #9615: [Core] Raylet port assignment race condition
### What is the problem?
Raylet is being assigned a port that is already in use, resulting in:
1.  A GRPC start failure that looks like: `server_chttp2.cc:40]        {"created":"@1588127532.732848116","description":"No address added out of total 1 resolved"...`
2. A failure to connect go Raylet by the RayletClient `raylet_client.cc:69] Retrying to connect to socket for pathname...`

The reason for this is that we assign a port in [`_get_unused_port`](https://github.com/ray-project/ray/blob/1fa305cf8b048361f9ebc5eeb30367877ae0e39d/python/ray/node.py#L427) and then use it in [`node_manager_server_.Run()`](https://github.com/ray-project/ray/blob/1fa305cf8b048361f9ebc5eeb30367877ae0e39d/src/ray/raylet/node_manager.cc#L228). There have been some attempts to 'hold onto the port in Python', but these do not work in the long run. 

This bug _**primarily comes up under artificial test conditions**_ where there is both a lot of node churn and many nodes are simulated on one computer. 

Causes  #8254
*Ray version and other system information (Python version, TensorFlow version, OS):*

### Reproduction (REQUIRED)
Please provide a script that can be run to reproduce the issue. The script should have **no external library dependencies** (i.e., use fake or mock data / environments):

If we cannot run your script, we cannot fix your issue.

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/latest/installation.html).

### PR DESCRIPTION
**Pull request by kfstorm**
## Why are these changes needed?

Currently, the node manager port is randomly generated in node.py. There's a slight chance that the random port is occupied by another process when Raylet is trying to start the node manager server with the port.

This PR fixes the issue by removing the randomized port in node.py, instead, we let Raylet pick the port. Users are still able to specify `--node-manager-port` manually in `ray start`.

This PR also helps stabilize Java UT, because the most common failure of Java UT is node manager port conflict.

TODO (I'm not going to address these in this PR):
- [ ] C++ worker get node manager port via `ray._private.services.get_address_info_from_redis`.
- [ ] The logic in `ray._private.services.get_address_info_from_redis` should be put into core worker.
- [ ] Raylet passes socket names and node manager port to worker processes via env variable, so we can remove `raylet_socket`, `store_socket`, and `node_manager_port` from `CoreWorkerOptions`.

## Related issue number

Closes #9615

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by kfstorm** [comment]
@rkooo567 What do you think about this PR?

**Comment by rkooo567** [comment]
Is it to avoid the port conflict?

**Comment by kfstorm** [comment]
Yes.

**Review by ericl — APPROVED** [review]

File: src/ray/raylet/worker_pool.cc [hunk 1]
```diff
@@ -217,6 +223,15 @@ Process WorkerPool::StartWorkerProcess(
...
     }
+    auto node_manager_port_position = token.find(kNodeManagerPortPlaceholder);
+    if (node_manager_port_position != std::string::npos) {
+      auto replaced_token = token;
+      replaced_token.replace(node_manager_port_position,
+                             strlen(kNodeManagerPortPlaceholder),
+                             std::to_string(node_manager_port_));
+      worker_command_args.push_back(replaced_token);
+      continue;
+    }
```
Comment (ijrsvt) [inline src/ray/raylet/worker_pool.cc:235]: Will this method always be called after SetNodeManagerPort?

**Comment by ijrsvt** [comment]
@kfstorm Thank you so much for finally closing this issue!!!!!

File: src/ray/raylet/worker_pool.cc (hunk 1 above)
Comment (kfstorm) [reply src/ray/raylet/worker_pool.cc:235]: Currently yes. But good catch! I should add `RAY_CHECK(node_manager_port_ != 0)` here.

Comment (ijrsvt) [reply src/ray/raylet/worker_pool.cc]: Thanks for adding that!

**Review by ijrsvt — APPROVED** [review]
Looks good!

**Pull request merged**